[SPMD] Introduce schedule_comm_wait

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98580
* #98579
* __->__ #98578
* #98285
* #98284
* #98182

`schedule_comm_wait` delays the wait_tensor ops as late as possible. Note that this optimization currently does not reorder the computation ops. For `foreach` based optimizer, we observe that reordering the computation ops is required to achieve a good performance.

Differential Revision: [D44761487](https://our.internmc.facebook.com/intern/diff/D44761487/)

fegin committed:
[SPMD] Introduce schedule_comm_wait

`schedule_comm_wait` delays the wait_tensor ops as late as possible. Note that this optimization currently does not reorder the computation ops. For `foreach` based optimizer, we observe that reordering the computation ops is required to achieve a good performance.

Differential Revision: [D44761487](https://our.internmc.facebook.com/intern/diff/D44761487/)

[ghstack-poisoned]

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -443,3 +443,41 @@ def comm_fusion_with_concat(
+    comm_blocks = get_all_comm_blocks(gm, (CommType.ALLREDUCE, "all_reduce"))
+
+    # Find all the end users.
+    allreduce_users = set()
```
> @mrshenli's internal review: Would I be correct that this optimization will be generalized to other collectives as well?

This pass only looks check the WaitTensor and reorder WaitTensor. So it should be generalized to other collectives  as well.

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -443,3 +443,41 @@ def comm_fusion_with_concat(
+            len(allreduce.outputs) >= 1
+        ), f"Found a allreduce that has zero outputs/users -- {allreduce}."
+        # Initialize the target_node to be the first user of the first output.
+        target_node = next(iter(next(iter(allreduce.outputs)).users))
```
> @mrshenli's internal review:  I guess this is assuming none of these users node will be reordered by subsequent graph optimization passes?

IMO, this is a general question for all optimization passes. If some subsequent optimization passes change the topological sort order, there is no guarantee that the previous optimization passes will still have the same effects, regardless the algorithm the previous passes adopt. I added `run_after` argument to the `graph_optimization_pass` contract to expect developers to consider the effect and add the order constraints to the graph optimization passes. But we may need a further discussion on this topic.

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -443,3 +443,41 @@ def comm_fusion_with_concat(
+                target_node = user
+                target_node_index = index
+
+        # Move wait nodes and all the subsequent output nodes before the
```
> @mrshenli's internal review:  We definitely need this, but I assume this won't help current dhen use case that much, because those allreduces are schedule right before the first usage in optimizer?

No, the expansion generates allreduce together with WaitTensor. So, without this pass, none of the allreduce will be overlapped with the computation. This was observed in the old _SPMD style expansion for DHEN and the new tracing with a simple MLP model. So I believe we need this pass for DHEN.

fegin replied:
`comm_fusion_with_concat` actually moves the allreduce to gradient producers. And this PR move the wait to the gradient consumers. So without reordering the computation op and memory constraint, this is the optimal scheduling.

fegin committed:
Update on "[SPMD] Introduce schedule_comm_wait"


`schedule_comm_wait` delays the wait_tensor ops as late as possible. Note that this optimization currently does not reorder the computation ops. For `foreach` based optimizer, we observe that reordering the computation ops is required to achieve a good performance.

Differential Revision: [D44761487](https://our.internmc.facebook.com/intern/diff/D44761487/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce schedule_comm_wait"


`schedule_comm_wait` delays the wait_tensor ops as late as possible. Note that this optimization currently does not reorder the computation ops. For `foreach` based optimizer, we observe that reordering the computation ops is required to achieve a good performance.

Differential Revision: [D44761487](https://our.internmc.facebook.com/intern/diff/D44761487/)

[ghstack-poisoned]

mrshenli reviewed (approved):
stamp to unblock. let's test accuracy parity before landing

fegin committed:
Update on "[SPMD] Introduce schedule_comm_wait"


`schedule_comm_wait` delays the wait_tensor ops as late as possible. Note that this optimization currently does not reorder the computation ops. For `foreach` based optimizer, we observe that reordering the computation ops is required to achieve a good performance.

Differential Revision: [D44761487](https://our.internmc.facebook.com/intern/diff/D44761487/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce schedule_comm_wait"


`schedule_comm_wait` delays the wait_tensor ops as late as possible. Note that this optimization currently does not reorder the computation ops. For `foreach` based optimizer, we observe that reordering the computation ops is required to achieve a good performance.

Differential Revision: [D44761487](https://our.internmc.facebook.com/intern/diff/D44761487/)

[ghstack-poisoned]

fegin commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -134,28 +138,56 @@ def train_step(model, optim, batch):
+        )
+        for _ in range(num_iters):
+            batch = torch.randn(batch_size, dim).cuda()
+            out = train_step(model, optim, batch)
```
Done
